sdk/go/common/util/deepcopy/copy.go

```diff
@@ -0,0 +1,88 @@
+package deepcopy
+
+import "reflect"
+
+// Copy returns a deep copy of the provided value.
+//
+// If there are multiple references to the same value inside the provided value, the multiply-referenced value will be
+// copied multiple times.
+func Copy(i interface{}) interface{} {
+	if i == nil {
+		return nil
+	}
+	return copy(reflect.ValueOf(i)).Interface()
+}
+
+func copy(v reflect.Value) reflect.Value {
+	if !v.IsValid() {
+		return v
+	}
+
+	typ := v.Type()
+	switch typ.Kind() {
+	case reflect.Bool,
+		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr,
+		reflect.Float32, reflect.Float64,
+		reflect.Complex64, reflect.Complex128,
+		reflect.String,
+		reflect.Func:
+		// These all have value semantics. Return them as-is.
+		return v
+	case reflect.Chan:
+		// Channels have referential semantics, but deep-copying them has no meaning. Return them as-is.
+		return v
+	case reflect.Interface:
+		rv := reflect.New(typ).Elem()
+		if !v.IsNil() {
+			rv.Set(copy(v.Elem()))
+		}
+		return rv
+	case reflect.Ptr:
+		if v.IsNil() {
+			return reflect.New(typ).Elem()
+		}
+		elem := copy(v.Elem())
+		if elem.CanAddr() {
+			return elem.Addr()
+		}
+		rv := reflect.New(typ.Elem())
+		rv.Set(elem)
+		return rv
+	case reflect.Array:
+		rv := reflect.New(typ).Elem()
+		for i := 0; i < v.Len(); i++ {
+			rv.Index(i).Set(copy(v.Index(i)))
```

Comment:
Would you mind writing some unit tests for these `copy` functions? It shouldn't be too cumbersome to create a prototypical "table driven" Golang unit test.

The reason I ask is that I'm not sure if this is correct. (And scanning [the docs](https://golang.org/pkg/reflect/#Value.Index), it isn't super clear.)

So we are creating a new "array type" and then copying every element over individually, great. But does the type information (`typ`) also include the length of the array, too? Or do we need to pre-allocate the length of `rv`, similar to line 61 for the slice-case?